eoe/man/man1/tput.1
'\"macro stdmacro
.if n .pH g1.tput @(#)tput	41.4 of 5/26/91
.\" Copyright 1991 UNIX System Laboratories, Inc.
.\" Copyright 1989, 1990 AT&T
.nr X
.if \nX=0 .ds x} tput 1 "Terminal Information Utilities" "\&"
.if \nX=1 .ds x} tput 1 "Terminal Information Utilities"
.if \nX=2 .ds x} tput 1 "" "\&"
.if \nX=3 .ds x} tput "" "" "\&"
.TH \*(x}
.SH NAME
\f4tput\f1 \- initialize a terminal or query terminfo database
.SH "SYNOPSIS"
\f4tput\f1 [\f4\-T\f2type\f1] \f2capname\f1 [\f2parms ... \f1]
.sp .5
\f4tput\f1
[\f4-T\f2type\f1]\f4 init\f1
.sp .5
\f4tput\f1
\f1[\f4-T\f2type\f1]\f4 reset\f1
.sp .5
\f4tput\f1
[\f4-T\f2type\f1]\f4 longname\f1
.sp .5
\f4tput -S  <<\f1
.SH "DESCRIPTION"
\f4tput\fP
uses the
\f4terminfo\fP
database to make the values of
terminal-dependent capabilities and information available to
the shell
(see \f4sh\fP(1)), to initialize or reset the terminal,
or return the long name of the requested terminal type.
\f4tput\fP
outputs a string if the attribute
.RI ( cap ability " name" )
is of type string, or an integer if the attribute is of type integer.
If the attribute is of type boolean, \f4tput\fP simply sets the exit code
(\f40\f1 for \s-1TRUE\s+1
if the terminal has the capability,
\f41\f1 for \s-1FALSE\s+1 if it does not),
and produces no output.
Before using a value returned on standard output,
the user should test the exit code [\f4$?\f1, see \f4sh\fP(1)]
to be sure it is \f40\f1.
(See the
.SM EXIT CODES
and
.SM DIAGNOSTICS
sections.)
For a complete list of capabilities and the
\f2capname\f1 associated with each, see \f4terminfo\fP(4).
.PP
.TP 11
\f4\-T\f2type\f1
indicates the \f2type\f1 of terminal.
Normally this option is unnecessary,
because the default is taken from the
environment variable
\f4TERM\f1.
If \f4-T\f1 is specified, then the shell variables
\f4LINES\f1 and \f4COLUMNS\f1
and the layer size [see \f4layers\fP(1)] will
not be referenced.
.TP
.I capname
indicates the attribute from the
\f4terminfo\fP
database.
.TP
.I parms
If the attribute is a string that takes parameters, the arguments
.I parms
will be instantiated into the string.
An all numeric argument will be passed to the attribute as a number.
.TP
\f4\-S\f1
allows more than one capability per invocation of \f4tput\f1.
The capabilities must be passed to \f4tput\f1 from the standard
input instead of from the command line (see example).
Only one \f2capname\f1 is allowed per line.
The \f4\-S\f1 option changes the meaning of the \f40\f1 and \f41\f1
boolean and string exit codes (see the
.SM EXIT CODES
section).
.TP
\f4init\f1
If the \f4terminfo\fP database is present
and an entry for the user's terminal
exists (see
\f4\-T\f2type,\f1
above), the following will occur:
(1) if present, the terminal's initialization
strings will be output
\f1(\f4is1\f1,
\f4is2\f1,
\f4is3\f1,
\f4if\f1,
\f4iprog\f1),
(2) any delays (e.g., newline)
specified in the entry will be set in the tty driver,
(3) tabs expansion will be turned on or
off according to the specification in the
entry, and (4) if tabs are not expanded,
standard tabs will be set (every 8 spaces).
If an entry does not contain the information
needed for any of the four above activities,
that activity will silently be skipped.
.TP
\f4reset\f1
Instead of putting out initialization strings,
the terminal's reset strings will be output
if present
\f1(\f4rs1\f1,
\f4rs2\f1,
\f4rs3\f1,
\f4rf\f1).
If the reset strings are not present,
but initialization strings are, the initialization
strings will be output.
Otherwise, \f4reset\f1 acts identically to \f4init\f1.
.br
.ne 10
.TP
\f4longname\f1
If the
\f4terminfo\fP
database is present and an entry for the user's
terminal exists
(see
\f4\-T\f2type\f1
above),
then the long name of the terminal will be put out.
The long name is the last name in the first line of
the terminal's description in the
\f4terminfo\fP
database
[see \f4term\fP(5)].
.SS International Functions
\f2capnames\fP for international functionality can also be specified.
.SH EXAMPLES
.TP 18
\f4tput init\f1
Initialize the terminal according to the
type of terminal in the environmental variable
\f4TERM\f1.
This command should be included in everyone's
\&.profile after the environmental variable \f4TERM\f1
has been exported, as illustrated on the \f4profile\fP(4)
manual page.
.TP
\f4tput \-T5620 reset\f1
Reset an AT&T 5620 terminal,
overriding the type of terminal in
the environmental variable
\f4TERM\f1.
.TP
\f4tput cup 0 0\f1
Send the sequence to move the cursor to row \f40\f1,
column \f40\f1 (the upper left corner of the screen,
usually known as the "home" cursor position).
.TP
\f4tput clear\f1
Echo the clear-screen sequence for the current terminal.
.TP
\f4tput cols\f1
Print the number of columns for the current terminal.
.TP
\f4tput -T450 cols\f1
Print the number of columns for the 450 terminal.
.TP
\f4bold=\f4`\fPtput smso\f4`\fP\f1
.TP
\f4offbold=\f4`\fPtput rmso\f4`\fP\f1
Set the shell variables \f4bold\f1,
to begin stand-out mode sequence,
and \f4offbold\f1, to end standout mode sequence,
for the current terminal.
This might be followed by a prompt:
.sp 0
\f4echo "${bold}Please type in your name: ${offbold}\\c"\f1
.TP
\f4tput hc\f1
Set exit code to indicate if the current terminal is a hardcopy terminal.
.TP
\f4tput cup 23 4\f1
Send the sequence to move the cursor to row 23, column 4.
.TP
\f4tput longname\f1
Print the long name from the
\f4terminfo\fP
database for the type of terminal
specified in the environmental variable
\f4TERM\f1.
.br
.ne 1.25i
.PD 0
.TP
\f4tput -S <<!\f1
.TP
\f4> clear\f1
.TP
\f4> cup 10 10\f1
.TP
\f4> bold\f1
.TP
\f4> !\f1
.sp -5
This example shows tput processing several capabilities in one invocation.
This example clears the screen, moves the cursor to position 10, 10
and turns on bold (extra bright) mode.
The list is terminated by an exclamation mark (\f4!\f1)
on a line by itself.
.PD
.SH FILES
.PD 0
.TP 2i
\f4/usr/share/lib/terminfo/?/\(**\f1
compiled terminal description database
.TP 2i
\f4/usr/include/curses.h\f1
\f4curses\fP(3X) header file
.TP 2i
\f4/usr/include/term.h\f1
\f4terminfo\fP header file
.TP 2i
\f4/usr/lib/tabset/\(**\f1
tab settings for some terminals, in a format appropriate to be
output to the terminal (escape sequences that set margins and tabs);
for more information, see the "Tabs and Initialization" section of
\f4terminfo\fP(4)
.PD
.SH "SEE ALSO"
\f4clear\fP(1),
\f4stty\fP(1),
\f4tabs\fP(1)
.br
\f4profile\fP(4), \f4terminfo\fP(4)
.SH EXIT CODES
If \f2capname\f1 is of type boolean,
a value of \f40\f1 is set for
\s-1TRUE\s0 and \f41\f1 for \s-1FALSE\s0
unless the \f4\-S\f1 option is used.
.PP
If \f2capname\f1 is of type string,
a value of \f40\f1 is set if the
.I capname
is defined for this terminal
.I type
(the value of
.I capname
is returned on standard output);
a value of
\f41\f1
is set if
.I capname
is not defined for this terminal
.I type
(a null value is returned on standard output).
.PP
If \f2capname\f1 is of type boolean or string and the \f4\-S\f1
option is used,
a value of \f40\f1 is returned to indicate
that all lines were successful.
No indication of which line failed can be given so exit code \f41\f1
will never appear.
Exit codes \f42\f1, \f43\f1, and \f44\f1 retain their usual interpretation.
.PP
If
.I capname
is of type integer,
a value of
\f40\f1
is always set,
whether or not
.I capname
is defined for this terminal
.IR type .
To determine if
.I capname
is defined for this terminal
.IR type ,
the user must test the value of standard output.
A value of
\f4\-1\f1
means that
.I capname
is not defined for this terminal
.IR type .
.P
Any other exit code indicates an error;
see the
.SM DIAGNOSTICS
section.
.SH DIAGNOSTICS
\f4tput\fP
prints the following error messages and
sets the corresponding exit codes.
.RS 3
.ta .5i
.nf
exit
code	\0\0\0\0\0\0\0\0\0\0\0\0error message
.sp .5
\0\f40\f1	\-1  (\f2capname\f1 is a numeric variable that is not specified in the
	\f4terminfo\fP(4) database for this terminal type, e.g.
	\f4tput \-T450 lines\f1 and \f4tput \-T2621 xmc\f1)
\0\f41\f1	no error message is printed, see the \f3\s-1EXIT CODES\s0\f1 section.
\0\f42\f1	usage error
\0\f43\f1	unknown terminal \f2type\f1 or no \f4terminfo\fP database
\0\f44\f1	unknown \f4terminfo\fP capability \f2capname\f1
.RE


eoe/lib/libcurses/screen/src/tput.c
/*	Copyright (c) 1990, 1991 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1988 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)curses:screen/tput.c	1.23"

/* tput - print terminal attribute

   return codes for cammand line arguments:
	0	ok (if boolean capname -> TRUE)
	1	(for boolean capname -> FALSE)

   return codes for standard input arguments:
	0	ok (tput for all lines was successful)

   return codes for both cases:
	2	usage error
	3	bad terminal type given or no terminfo database
	4	unknown capname

   tput printfs (a value) if an INT capname was given (e.g. cols),
	putp's (a string) if a STRING capname was given (e.g. clear), and
	for BOOLEAN capnames (e.g. hard-copy) just returns the boolean value.
*/

#include "curses_inc.h"
#include <fcntl.h>
#include <ctype.h>

struct delay;
static void cat(char *);
static void initterm(void);
static void reset_term(void);
static void setdelay(int, struct delay [], int, tcflag_t *);
static int settabs(void);
static int allnumeric(char *);
static int getpad(char *);
static int outputcap(char *, int, char **);

char *progname;		/* argv[0] */
int CurrentBaudRate;	/* current baud rate */
int reset = 0;		/* called as reset_term */
int fildes = 1;

main(int argc, char **argv)
{
    register int i, std_argc;
    register char *term = getenv("TERM");
    register char *cap, std_input = FALSE;
    int setuperr;

    progname = argv[0];

    while ((i = getopt (argc, argv, "ST:")) != EOF)
	switch (i)
	    {
	    case 'T':
		fildes = -1;
		putenv("LINES=");
		putenv("COLUMNS=");
		term = optarg;
		break;
	    case 'S':
		std_input = TRUE;
		break;
	    case '?':
	    usage:
		(void) fprintf (stderr,
		    "usage: \"%s [-ST term] capname [tparm argument...]\"\n",
		    progname);
		exit(2);
	    }

    if (!term || !*term)
	{
	(void) fprintf(stderr,"%s: No value for $TERM and no -T specified\n",
	    progname);
	exit(2);
	}

    (void) setupterm (term, fildes, &setuperr);

    switch (setuperr)
	{
	case -2:
	    (void) fprintf(stderr,"%s: unreadable terminal descriptor \"%s\"\n",
		progname, term);
	    exit(3);
	case -1:
	    (void) fprintf(stderr,"%s: no terminfo database\n",
		progname);
	    exit(3);
	case 0:
	    (void) fprintf(stderr,"%s: unknown terminal \"%s\"\n",
		progname, term);
	    exit(3);
	}

    reset_shell_mode();

    /* command line arguments */
    if (!std_input)
    {
        if (argc == optind)
	    goto usage;
	cap = argv[optind++];

	if (strcmp(cap, "init") == 0)
	    initterm();
	else if (strcmp(cap, "reset") == 0)
	    reset_term();
	else if (strcmp(cap, "longname") == 0)
	    printf("%s\n",longname());
	else
	    exit (outputcap(cap, argc, argv));
	exit (0);
    }

    /* standard input argumets	*/
    else
    {
#define BUFLEN 128
	char buff[BUFLEN];
	register char **v;

	/* allocate storage for the 'faked' argv[] array	*/
	v = (char **) malloc (10 * sizeof (char *));
	for (i=0; i<10; i++)
	     v[i] = (char *) malloc (32 * sizeof (char));
	while (fgets (buff,BUFLEN,stdin) != NULL)
	{
	    /* read standart input line; skip over empty lines */
	    if ((std_argc = sscanf (buff, "%s %s %s %s %s %s %s %s %s %s",
		v[0], v[1], v[2], v[3], v[4], v[5], v[6], v[7], v[8], v[9])) < 1)
		continue;

	    cap = v[0];
	    optind = 1;

	    if (strcmp(cap, "init") == 0)
		initterm();
	    else if (strcmp(cap, "reset") == 0)
		reset_term();
	    else if (strcmp(cap, "longname") == 0)
		printf("%s\n",longname());
	    else
		outputcap(cap, std_argc, v);
	    fflush (stdout);
	}
	exit (0);
    }
}

static long parm[9] =
{
    0, 0, 0, 0, 0, 0, 0, 0, 0
};

static int
outputcap(register char *cap, int argc, char **argv)
{
    register int parmset = 0;
    char *thisstr;
    register int i;

    if ((i = tigetflag(cap)) >= 0)
	return(1 - i);

    if ((i = tigetnum(cap)) >= -1)
	{
	(void) printf ("%d\n", i);
	return (0);
	}

    if ((thisstr = tigetstr(cap)) != (char *)-1)
	{
	if (!thisstr) return(1);
	for (parmset = 0; optind < argc; optind++, parmset++)
	    if (allnumeric(argv[optind]))
		parm[parmset] = atoi(argv[optind]);
	    else
		parm[parmset] = (long) argv[optind];

	if (parmset)
	    putp (tparm (thisstr,
		parm[0], parm[1], parm[2], parm[3], parm[4], parm[5],
		parm[6], parm[7], parm[8]));
	else
	    putp (thisstr);
	return (0);
	}

    (void) fprintf(stderr,"%s: unknown terminfo capability '%s'\n", progname, cap);
    exit (4);
    return(0);
}

/*
    The decision as to whether an argument is a number or not is to simply
    look at whether there are any non-digits in the string.
*/
int
allnumeric(register char *string)
{
    if (*string)
    {
        while (*string)
	    if (!isdigit(*string++))
	        return 0;
        return 1;
    }
    else
	return 0;
}

/*
    SYSTEM DEPENDENT TERMINAL DELAY TABLES

	These tables maintain the correspondence between the delays
	defined in terminfo and the delay algorithms in the tty driver
	on the particular systems. For each type of delay, the bits used
	for that delay must be specified (in XXbits) and a table
	must be defined giving correspondences between delays and
	algorithms. Algorithms which are not fixed delays (such
	as dependent on current column or line number) must be
	kludged in some way at this time.

	Some of this was taken from tset(1).
*/

struct delay
{
    int d_delay;
    int d_bits;
};


#if defined(SYSV) || defined(USG)
/*	Unix 3.0 on up */

/*    Carriage Return delays	*/

int	CRbits = CRDLY;
struct delay	CRdelay[] =
{
	0,	CR0,
	80,	CR1,
	100,	CR2,
	150,	CR3,
	-1
};

/*	New Line delays	*/

int	NLbits = NLDLY;
struct delay	NLdelay[] =
{
	0,	NL0,
	100,	NL1,
	-1
};

/*	Back Space delays	*/

int	BSbits = BSDLY;
struct delay	BSdelay[] =
{
	0,	BS0,
	50,	BS1,
	-1
};

/*	TaB delays	*/

int	TBbits = TABDLY;
struct delay	TBdelay[] =
{
	0,	TAB0,
	11,	TAB1,		/* special M37 delay */
	100,	TAB2,
				/* TAB3 is XTABS and not a delay */
	-1
};

/*	Form Feed delays	*/

int	FFbits = FFDLY;
struct delay	FFdelay[] =
{
	0,	FF0,
	2000,	FF1,
	-1
};

#else	/* BSD */

/*	Carriage Return delays	*/

int	CRbits = CRDELAY;
struct delay	CRdelay[] =
{
	0,	CR0,
	9,	CR3,
	80,	CR1,
	160,	CR2,
	-1
};

/*	New Line delays	*/

int	NLbits = NLDELAY;
struct delay	NLdelay[] =
{
	0,	NL0,
	66,	NL1,		/* special M37 delay */
	100,	NL2,
	-1
};

/*	Tab delays	*/

int	TBbits = TBDELAY;
struct delay	TBdelay[] =
{
	0,	TAB0,
	11,	TAB1,		/* special M37 delay */
	-1
};

/*	Form Feed delays	*/

int	FFbits = VTDELAY;
struct delay	FFdelay[] =
{
	0,	FF0,
	2000,	FF1,
	-1
};
#endif	/* BSD */

/*
    Initterm (reset_term) does terminal specific initialization. In
    particular, the init_strings from terminfo are output and tabs are
    set (if they aren't hardwired in). Much of this stuff was done by
    the tset(1) program.
*/

/*
   Figure out how many milliseconds of padding the capability cap
   needs and return that number. Padding is stored in the string as $<n>,
   where n is the number of milliseconds of padding. More than one
   padding string is allowed within the string, although this is unlikely.
*/

static int
getpad (register char *cap)
{
    register int padding = 0;

    /* No padding needed at speeds below padding_baud_rate */
    if ((padding_baud_rate > CurrentBaudRate) || ( cap == NULL ))
	return 0;

    while (*cap)
	if ( (cap[0] == '$') && (cap[1] == '<') )
	    {
	    cap++;
	    cap++;
	    padding += atoi (cap);
	    while ( isdigit (*cap) )
		cap++;
	    while (*cap == '.' || *cap == '/' || *cap == '*' || isdigit(*cap))
		cap++;
	    while (*cap == '>')
		cap++;
	    }
	else
	    cap++;
    return padding;
}

/*
    Set the appropriate delay bits in the termio structure for
    the given delay.
*/
static void
setdelay (register int delay, struct delay delaytable[], int bits, tcflag_t *flags)
{
    register struct delay  *p;
    register struct delay  *lastdelay;

    /* Clear out the bits, replace with new ones */
    *flags &= (tcflag_t) ~bits;

    /* Scan the delay table for first entry with adequate delay */
    for (lastdelay = p = delaytable;
	 (p -> d_delay >= 0) && (p -> d_delay < delay);
	 p++)
	lastdelay = p;

    /* use last entry if none will do */
    *flags |= (tcflag_t) lastdelay -> d_bits;
}

/*
 * Set the hardware tabs on the terminal, using clear_all_tabs,
 * set_tab, and column_address capabilities. Cursor_address and cursor_right
 * may also be used, if necessary.
 * This is done before the init_file and init_3string, so they can patch in
 * case we blow this.
 */

static int
settabs(void)
{
    register int c;

    /* Do not set tabs if they power up properly. */
    if (init_tabs == 8)
	return (0);

    if (set_tab)
	{
	/* Force the cursor to be at the left margin. */
	if (carriage_return)
	    putp (carriage_return);
	else
	    (void) putchar ('\r');

	/* Clear any current tab settings. */
	if (clear_all_tabs)
	    putp (clear_all_tabs);

	/* Set the tabs. */
	for (c = 8; c < columns; c += 8)
	    {
	    /* Get to that column. */
	    (void) fputs ("        ", stdout);

	    /* Set the tab. */
	    putp (set_tab);
	    }

	/* Get back to the left column. */
	if (carriage_return)
	    putp (carriage_return);
	else
	    (void) putchar ('\r');

	return 1;
	}
    return 0;
}

/*
    Copy "file" onto standard output.
*/

static void
cat (char *file)
     /* file = File to copy. */
{
    register int fd;			/* File descriptor. */
    register int i;			/* Number characters read. */
    char buf[BUFSIZ];			/* Buffer to read into. */

    fd = open (file, O_RDONLY);
    if (fd < 0)
	(void) fprintf (stderr, "%s: Cannot open initialization file %s.\n",
	    progname, file);
    else
	{
	while ((i = read (fd, buf, BUFSIZ)) > 0)
	    (void) write (fileno (stdout), buf, (unsigned) i);
	(void) close (fd);
	}
}

/*
    Initialize the terminal.
    Send the initialization strings to the terminal.
*/

static void
initterm (void)
{
    register int filedes;		/* File descriptor for ioctl's. */
#if defined(SYSV) || defined(USG)
# if _ABIAPI
    struct termios termmode;		/* To hold terminal settings. */
# define GTTY(fd,mode)	ioctl(fd, TCGETS, mode)
# define STTY(fd,mode)	ioctl(fd, TCSETSW, mode)
# else
    struct termio termmode;		/* To hold terminal settings. */
# define GTTY(fd,mode)	ioctl(fd, TCGETA, mode)
# define STTY(fd,mode)	ioctl(fd, TCSETAW, mode)
# endif
# define SPEED(mode)	_BR(mode)
# define OFLAG(mode)	mode.c_oflag
#else	/* BSD */
    struct sgttyb termmode;		/* To hold terminal settings. */
# define GTTY(fd,mode)	gtty(fd, mode)
# define STTY(fd,mode)	stty(fd, mode)
# define SPEED(mode)	(mode.sg_ospeed & 017)
# define OFLAG(mode)	mode.sg_flags
# define TAB3		XTABS
#endif

    /* Get the terminal settings. */
    /* First try standard output, then standard error, */
    /* then standard input, then /dev/tty. */
    if ( (filedes = 1, GTTY (filedes, &termmode) == -1) ||
	 (filedes = 2, GTTY (filedes, &termmode) == -1) ||
	 (filedes = 0, GTTY (filedes, &termmode) == -1) ||
	 (filedes = open ("/dev/tty", O_RDWR),
	     GTTY (filedes, &termmode) == -1) )
	{
	filedes = -1;
	CurrentBaudRate = 1200;
	}
    else
	CurrentBaudRate = SPEED(termmode);

    if (xon_xoff)
	{
	OFLAG(termmode) &= (unsigned) ~(NLbits | CRbits | BSbits | FFbits | TBbits);
	}
    else
	{
	setdelay(getpad(carriage_return), CRdelay, CRbits, &OFLAG(termmode));
	setdelay(getpad(scroll_forward), NLdelay, NLbits, &OFLAG(termmode));
	setdelay(getpad(cursor_left), BSdelay, BSbits, &OFLAG(termmode));
	setdelay(getpad(form_feed), FFdelay, FFbits, &OFLAG(termmode));
	setdelay(getpad(tab), TBdelay, TBbits, &OFLAG(termmode));
	}

    /* If tabs can be sent to the tty, turn off their expansion. */
    if (tab && set_tab || init_tabs == 8)
	OFLAG(termmode) &= ~(TAB3);
    else
	OFLAG(termmode) |= TAB3;

    /* Do the changes to the terminal settings */
    (void) STTY (filedes, &termmode);

    /* Send first initialization strings. */
    if (init_prog)
	(void) system (init_prog);

    if (reset && reset_1string)
	putp (reset_1string);
    else if (init_1string)
	putp (init_1string);

    if (reset && reset_2string)
	putp (reset_2string);
    else if (init_2string)
	putp (init_2string);

    /* Set up the tabs stops. */
    settabs ();

    /* Send out initializing file. */
    if (reset && reset_file)
	cat (reset_file);
    else if (init_file)
	cat (init_file);

    /* Send final initialization strings. */
    if (reset && reset_3string)
	putp (reset_3string);
    else if (init_3string)
	putp (init_3string);

    if (carriage_return)
        putp (carriage_return);
    else
	(void) putchar ('\r');

    /* Send color initialization strings */

    if (orig_colors)
	putp (orig_colors);

    if (orig_pair)
	putp ( tparm(orig_pair) );

    /* Let the terminal settle down. */
    (void) fflush (stdout);
    (void) sleep (1);
}

static void
reset_term(void)
{
    reset++;
    initterm();
}
